firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHTTPClient.java

```diff
@@ -0,0 +1,233 @@
+import java.util.Random;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+import okhttp3.ConnectionPool;
+import okhttp3.OkHttpClient;
+import okhttp3.Protocol;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.sse.EventSource;
+import okhttp3.sse.EventSourceListener;
+import okhttp3.sse.EventSources;
+
+public class ConfigRealtimeHTTPClient {
+    // Headers
+    private static final String API_KEY_HEADER = "X-Goog-Api-Key";
+    private static final String X_ANDROID_PACKAGE_HEADER = "X-Android-Package";
+    private static final String X_ANDROID_CERT_HEADER = "X-Android-Cert";
+    private static final String X_GOOGLE_GFE_CAN_RETRY = "X-Google-GFE-Can-Retry";
+    private static final String INSTALLATIONS_AUTH_TOKEN_HEADER =
+            "X-Goog-Firebase-Installations-Auth";
+    
+    private static final String REALTIME_URL_STRING = "http://10.0.2.2:5000/sse";
+    private static final Logger logger = Logger.getLogger("Real_Time_RC");
+
+    private final ConfigFetchHandler configFetchHandler;
+    private final FirebaseApp firebaseApp;
+    private final FirebaseInstallationsApi firebaseInstallations;
+    private final Executor executor;
+    private final Context context;
+
+    // Retry parameters
+    private final Timer timer;
+    private final int ORIGINAL_RETRIES = 7;
+    private final long RETRY_TIME = 10000;
+    private long RETRY_MULTIPLIER;
+    private int RETRIES_REMAINING;
+    private final Random random;
+
+    // HTTP/2 client and SSE EventSource
+    private final OkHttpClient okHttpClient;
+    private EventSource eventSource;
+
+    // Map of callbacks
+    private final Map<String, RealTimeEventListener> eventListeners;
+
+    public ConfigRealtimeHTTPClient(FirebaseApp firebaseApp,
+                                    FirebaseInstallationsApi firebaseInstallations,
+                                    ConfigFetchHandler configFetchHandler,
+                                    Context context,
+                                    Executor executor) {
+        this.firebaseApp = firebaseApp;
+        this.configFetchHandler = configFetchHandler;
+        this.firebaseInstallations = firebaseInstallations;
+        this.context = context;
+        this.executor = executor;
+        this.eventListeners = new HashMap<>();
+
+        // Retry parameters
+        this.random = new Random();
+        this.timer = new Timer();
+
+        // HTTP Client
+        ConnectionPool connectionPool = new ConnectionPool();
+        this.okHttpClient = new OkHttpClient.Builder()
+                .readTimeout(2, TimeUnit.HOURS)
+                .writeTimeout(2, TimeUnit.HOURS)
+                .connectTimeout(2, TimeUnit.HOURS)
+                .followRedirects(false)
+                .protocols(Collections.singletonList(Protocol.H2_PRIOR_KNOWLEDGE))
+                .retryOnConnectionFailure(true)
```

Comment:
Do we know the semantics of this retry functionality when used in conjunction with okhttp-sse? also, given the 2 hour timeouts above, will we retry after that long?

Reply:
We need to ensure that we use backoff for these retries.

Reply:
I was planning on relying more on the retry function that I made that also includes random backoffs, configurable amount of retries, and async scheduling.